src/base/test/test_string.cc

```diff
@@ -115,4 +115,65 @@ TESTCASE(trim_split_replace) {
 	check_equal(str, "foo/wordr/wordz");
 }
 
+TESTCASE(string_formatting) {
+	check_equal("Hello World", bformat("%s", "Hello World"));
+	check_equal("Hello World", bformat("%s %s", "Hello", "World"));
+	check_equal("Hello World", bformat("%1$s %2%", "Hello", "World"));
+	check_equal("Hello World", bformat("%2% %1%", "World", "Hello"));
+	check_equal("   Hello World", bformat("%1$14s", "Hello World"));
+	check_equal("Hello World   ", bformat("%-14s", "Hello World"));
+	check_equal("         Hello", bformat("%14.5s", "Hello World"));
+	check_equal("Hello         ", bformat("%1$-14.5s", "Hello World"));
+
+	check_equal("A123X", bformat("A%dX", 123));
+	check_equal("AABCDEFX", bformat("A%XX", 0xABCDEF));
+	check_equal("A-1X", bformat("A%dX", -1));
+	check_equal("A+123X", bformat("A%0+2dX", 123));
+	check_equal("A+123    X", bformat("A%+-8dX", 123));
+	check_equal("A+123    X", bformat("A%-+8dX", 123));
+	check_equal("A123     X", bformat("A%-8dX", 123));
+	check_equal("A     123X", bformat("A%8dX", 123));
+	check_equal("A00000123X", bformat("A%08dX", 123));
+	check_equal("A    +123X", bformat("A%+8dX", 123));
+	check_equal("A0000+123X", bformat("A%+08dX", 123));
```

Comment:
Isn't that supposed to be `"A+0000123X"`? Also for negative numbers without the `+` flag. That will need one more test to change the order of writing the padding and the sign depending on the `0` flag.